packages/block-editor/src/components/font-appearance-control/index.js

```diff
@@ -98,12 +105,46 @@ export default function FontAppearanceControl( props ) {
 		style: { fontStyle: undefined, fontWeight: undefined },
 	};
 
+	// Build font style and weight options based on the font family faces.
+	let fontWeights = [];
+	let fontStyles = [];
+	let variableFont = false;
+
+	fontFamilyFaces?.forEach( ( face ) => {
+		if ( face.fontWeight ) {
+			if (
+				! fontWeights.some(
+					( weight ) => weight.value === face.fontWeight
+				)
+			) {
+				// Check if font weight includes a space, if so it must be a variable font.
+				if ( /\s/.test( face.fontWeight.trim() ) ) {
+					variableFont = true;
+				}
+
+				fontWeights.push( formatFontWeight( face.fontWeight ) );
+			}
+		}
+		if ( face.fontStyle ) {
+			if (
+				! fontStyles.some( ( style ) => style.value === face.fontStyle )
+			) {
+				fontStyles.push( formatFontStyle( face.fontStyle ) );
+			}
+		}
+	} );
+
+	fontWeights =
+		fontWeights.length === 0 || variableFont ? FONT_WEIGHTS : fontWeights;
+	fontStyles =
+		fontStyles.length === 0 || variableFont ? FONT_STYLES : fontStyles;
+
```

Comment:
This logic is complex enough to be worth extracting as a pure function and adding test cases for it. The tests would also help to document the expected output for each theme.json definition. I think that's useful for the follow-up tasks regarding variable typographies, too.

Reply:
Makes sense! I've abstracted this to a function here: https://github.com/WordPress/gutenberg/pull/61915/commits/e6d2e67d0898b361c13fd092bae6497e7529b010.